qcodes/tests/dataset/test_database_extract_runs.py

```diff
@@ -0,0 +1,477 @@
+    # the source objects?
+
+    assert source_dataset.the_same_dataset_as(target_dataset)
+
+    source_data = source_dataset.get_data(*source_dataset.parameters.split(','))
+    target_data = target_dataset.get_data(*target_dataset.parameters.split(','))
+
+    assert source_data == target_data
+
+    exp_attrs = ['name', 'sample_name', 'format_string', 'started_at',
+                 'finished_at']
+
+    for exp_attr in exp_attrs:
+        assert getattr(source_exp, exp_attr) == getattr(target_exp, exp_attr)
+
+    # trying to insert the same run again should be a NOOP
+    with raise_if_file_changed(target_path):
+        extract_runs_into_db(source_path, target_path, source_dataset.run_id)
+
+
+def test_correct_experiment_routing(two_empty_temp_db_connections,
+                                    some_paramspecs):
+    """
+    Test that existing experiments are correctly identified AND that multiple
+    insertions of the same runs don't matter (run insertion is idempotent)
+    """
+    source_conn, target_conn = two_empty_temp_db_connections
+
+    source_exp_1 = Experiment(conn=source_conn)
+
+    # make 5 runs in first experiment
+
+    exp_1_run_ids = []
+    for _ in range(5):
+
+        source_dataset = DataSet(conn=source_conn, exp_id=source_exp_1.exp_id)
+        exp_1_run_ids.append(source_dataset.run_id)
+
+        for ps in some_paramspecs[2].values():
+            source_dataset.add_parameter(ps)
+
+        for val in range(10):
+            source_dataset.add_result({ps.name: val
+                                       for ps in some_paramspecs[2].values()})
+        source_dataset.mark_complete()
+
+    # make a new experiment with 1 run
+
+    source_exp_2 = Experiment(conn=source_conn)
+    ds = DataSet(conn=source_conn, exp_id=source_exp_2.exp_id, name="lala")
+    exp_2_run_ids = [ds.run_id]
+
+    for ps in some_paramspecs[2].values():
+        ds.add_parameter(ps)
+
+    for val in range(10):
+        ds.add_result({ps.name: val for ps in some_paramspecs[2].values()})
+
+    ds.mark_complete()
+
+    source_path = path_to_dbfile(source_conn)
+    target_path = path_to_dbfile(target_conn)
+
+    # now copy 2 runs
+    extract_runs_into_db(source_path, target_path, *exp_1_run_ids[:2])
+
+    target_exp1 = Experiment(conn=target_conn, exp_id=1)
+
+    assert len(target_exp1) == 2
+
+    # copy two other runs, one of them already in
+    extract_runs_into_db(source_path, target_path, *exp_1_run_ids[1:3])
+
+    assert len(target_exp1) == 3
+
+    # insert run from different experiment
+    extract_runs_into_db(source_path, target_path, ds.run_id)
+
+    assert len(target_exp1) == 3
+
+    target_exp2 = Experiment(conn=target_conn, exp_id=2)
+
+    assert len(target_exp2) == 1
+
+    # finally insert every single run from experiment 1
+
+    extract_runs_into_db(source_path, target_path, *exp_1_run_ids)
+
+    # check for idempotency once more by inserting all the runs but in another
```

Comment:
for rigidity of the test, it'll better if this "check for idempotency" is done not here, but in front of "# check that all the datasets match up" (right after "assert len(target_exp2) == 1" below)

Reply:
Sorry, I don't understand what you mean?

Reply:
I'm suggesting moving lines 211-214 into the space between lines 220 and 222 :)